File: app/schemas/comparison.py
from pydantic import BaseModel, Field, field_validator

from app.rules.mission_profiles import is_supported_task_type, normalize_task_type
from app.schemas.assessment import CruiseAssessmentAdvice, RiskDecision
from app.schemas.recommendation import RecommendationWindow


class MultiLocationComparisonRequest(BaseModel):
    locations: list[str] = Field(..., min_length=2, description="候选地点列表，至少 2 个")
    date: str
    start_time: str
    end_time: str
    task_type: str = "cruise"
    purpose: str | None = None
    top_k: int = 3
    comparison_mode: str = "default"

    @field_validator("task_type")
    @classmethod
    def validate_task_type(cls, value: str) -> str:
        normalized = normalize_task_type(value)
        if not is_supported_task_type(normalized):
            raise ValueError("task_type must be one of: cruise, inspection, hover, survey")
        return normalized


class ComparedLocationResult(BaseModel):
    rank: int
    location: str
    available: bool = True
    location_id: str | None = None
    overall_decision: RiskDecision | None = None
    allow_cruise: bool | None = None
    risk_score: int | None = None
    score: float | None = None
    flyable_hour_count: int = 0
    max_continuous_flyable_hours: int = 0
    earliest_flyable_time: str | None = None
    window_quality_score: float | None = None
    best_window: RecommendationWindow | None = None
    score_breakdown: dict[str, float] = Field(default_factory=dict)
    summary_risk_factors: list[str] = Field(default_factory=list)
    error_message: str | None = None
    advice: CruiseAssessmentAdvice | None = None


class MultiLocationComparisonResponse(BaseModel):
    request: dict[str, object]
    comparisons: list[ComparedLocationResult] = Field(default_factory=list)
    top_k_locations: list[ComparedLocationResult] = Field(default_factory=list)
    recommended_location: ComparedLocationResult | None = None

File: frontend/src/types/comparison.ts
import type { CruiseAssessmentAdvice, TaskType } from './evaluation'
import type { RecommendationWindow } from './recommendation'

export type MultiLocationComparisonRequest = {
  locations: string[]
  date: string
  start_time: string
  end_time: string
  task_type: TaskType
  purpose?: string | null
  top_k: number
  comparison_mode: string
}

export type ComparedLocationResult = {
  rank: number
  location: string
  available: boolean
  location_id?: string | null
  overall_decision?: string | null
  allow_cruise?: boolean | null
  risk_score?: number | null
  score?: number | null
  flyable_hour_count: number
  max_continuous_flyable_hours: number
  earliest_flyable_time?: string | null
  window_quality_score?: number | null
  best_window?: RecommendationWindow | null
  score_breakdown: Record<string, number>
  summary_risk_factors: string[]
  error_message?: string | null
  advice?: CruiseAssessmentAdvice | null
}

export type MultiLocationComparisonResponse = {
  request: Record<string, unknown>
  comparisons: ComparedLocationResult[]
  top_k_locations: ComparedLocationResult[]
  recommended_location?: ComparedLocationResult | null
}
